test/gurka/test_reachability.cc

```diff
@@ -0,0 +1,51 @@
+#include "gurka.h"
+#include <gtest/gtest.h>
+
+using namespace valhalla;
+
+TEST(Reachability, dont_snap_waypoints_to_deadends) {
+  const std::string ascii_map = R"(
+  A--1-B-2--C
+      / \
+     D   E)";
+
+  // BD and BE are oneways that lead away, and toward the main road A-B-C
+  const gurka::ways ways = {{"AB", {{"highway", "primary"}}},
+                            {"BC", {{"highway", "primary"}}},
+                            {"BD", {{"highway", "secondary"}, {"oneway", "yes"}}},
+                            {"BE", {{"highway", "secondary"}, {"oneway", "yes"}}}};
+  const auto layout = gurka::detail::map_to_coordinates(ascii_map, 25);
+
+  std::unordered_map<std::string, std::string> config_opts = {
+      {"mjolnir.concurrency", "1"},
+      {"loki.service_defaults.minimum_reachability", "3"},
+  };
+
+  auto map = gurka::buildtiles(layout, ways, {}, {}, "test/data/reachability", config_opts);
+
+  {
+    // Verify simple waypoint routing works
+    auto result1 = gurka::do_action(valhalla::Options::route, map, {"A", "1", "C"}, "auto", {}, {},
+                                    nullptr, "break_through");
+    gurka::assert::raw::expect_path(result1, {"AB", "AB", "AB", "BC"});
+    auto result2 = gurka::do_action(valhalla::Options::route, map, {"A", "2", "C"}, "auto", {}, {},
+                                    nullptr, "break_through");
+    gurka::assert::raw::expect_path(result2, {"AB", "BC", "BC"});
+  }
+  {
+    // BD is a oneway leading away from ABC with no escape
+    // input coordinate at D should get snapped to the ABC way
+    // Should give the same result as A->1->C
+    auto result = gurka::do_action(valhalla::Options::route, map, {"A", "D", "C"}, "auto", {}, {},
+                                   nullptr, "break_through");
+    gurka::assert::raw::expect_path(result, {"AB", "AB", "AB", "BC"});
```

Comment:
why are there three AB edges? im sure there is some detail i just cant remember at the moment... i get why there is one from A to the snap point at or near 1 and that then there is another for the next leg from that point to B but i cant figure out where the 3rd one is coming from at the moment. can you jog my memory?

Reply:
ha, great question!) actually, the first `AB` edge is reversed. I think there is a small bug in `TimeDepForward` algorithm. I debugged it, there is a combination of conditions in `TimeDepForward::SetOrigin` method that led to this:
1. looks like we should remove all inbound edges in case of the origin was snapped to the node, but we didn't do ti because of this https://github.com/valhalla/valhalla/blob/master/src/thor/timedep_forward.cc#L432 : origin and destination is on the same edge;
2. when we got forward `AB` edge we found that the destination also on this edge, and added forward `AB` edge to the priority queue with cost equal to `A -> 1`, BUT didn't settle this edge because of this comment https://github.com/valhalla/valhalla/blob/master/src/thor/timedep_forward.cc#L530 ;
3. last step, we have edges `ABr` and `ABf` in the queue and sortcost for `ABr` edge is less than sortcost for `ABf`. So, pop edge `ABr` from the queue, start expansion from the node `A` and found destination edge `ABf`. The end)

Reply:
I think, this should be fixed in another PR.

Reply:
interesting, we have a bit of code that is supposed to do this if i can remember where it is. it should say somethign like if its node snapped and its an origin no point in using the edge that is snapped to its end node...

Reply:
this bit of code: https://github.com/valhalla/valhalla/blob/master/src/thor/timedep_forward.cc#L424-L429

Reply:
the purpose is to ignore end node snaps for origins of the route unless that is all we have

Reply:
https://github.com/valhalla/valhalla/blob/master/src/thor/timedep_forward.cc#L432 - we also use this condition that just checks that origin and destination on the same edge. so , probably it's needless